File: backend/app/models/book.py
from sqlalchemy import Column, Integer, String, Boolean, DateTime, Text, JSON, BigInteger, ForeignKey, DECIMAL
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from .database import Base

class Book(Base):
    __tablename__ = "books"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    
    # Basic metadata
    title = Column(String(500), nullable=False)
    author = Column(String(200))
    language = Column(String(10), default="ja")
    
    # File information
    file_path = Column(String(500), nullable=False)
    file_name = Column(String(255), nullable=False)
    file_size = Column(BigInteger)
    mime_type = Column(String(100))
    
    # Processing status
    processing_status = Column(String(20), default="pending")  # pending, processing, completed, failed
    word_count = Column(Integer, default=0)
    unique_word_count = Column(Integer, default=0)
    difficulty_level = Column(String(10))  # beginner, intermediate, advanced
    
    # Timestamps
    uploaded_at = Column(DateTime(timezone=True), server_default=func.now())
    last_read_at = Column(DateTime(timezone=True))
    
    # Reading progress
    reading_progress = Column(DECIMAL(5, 2), default=0.00)  # percentage
    
    # Content extraction
    extracted_text = Column(Text)
    chapter_data = Column(JSON, default=list)  # [{title, start_pos, end_pos}]
    
    # Relationships
    user = relationship("User", back_populates="books")
    reading_sessions = relationship("ReadingSession", back_populates="book")
    annotations = relationship("BookAnnotation", back_populates="book")
    
    def __repr__(self):
        return f"<Book(title='{self.title}', author='{self.author}', user_id={self.user_id})>"
    
    def to_dict(self):
        """Convert book object to dictionary for API responses"""
        return {
            "id": self.id,
            "user_id": self.user_id,
            "title": self.title,
            "author": self.author,
            "language": self.language,
            "file_name": self.file_name,
            "file_size": self.file_size,
            "mime_type": self.mime_type,
            "processing_status": self.processing_status,
            "word_count": self.word_count,
            "unique_word_count": self.unique_word_count,
            "difficulty_level": self.difficulty_level,
            "uploaded_at": self.uploaded_at.isoformat() if self.uploaded_at else None,
            "last_read_at": self.last_read_at.isoformat() if self.last_read_at else None,
            "reading_progress": float(self.reading_progress) if self.reading_progress else 0.0,
            "chapter_data": self.chapter_data or []
        }

# Add the relationship to User model
from .user import User
User.books = relationship("Book", back_populates="user")

File: backend-go/internal/models/book.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// Book represents an uploaded ebook
type Book struct {
	ID        uint           `json:"id" gorm:"primarykey"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`

	// Foreign key
	UserID uint `json:"user_id" gorm:"not null;index"`
	User   User `json:"user,omitempty" gorm:"foreignKey:UserID"`

	// Basic metadata
	Title    string `json:"title" gorm:"size:500;not null"`
	Author   string `json:"author" gorm:"size:200"`
	Language string `json:"language" gorm:"size:10;default:ja"`

	// File information
	FilePath string `json:"file_path" gorm:"size:500;not null"`
	FileName string `json:"file_name" gorm:"size:255;not null"`
	FileSize int64  `json:"file_size"`
	MimeType string `json:"mime_type" gorm:"size:100"`

	// Processing status
	ProcessingStatus string `json:"processing_status" gorm:"size:20;default:pending"` // pending, processing, completed, failed
	WordCount        int    `json:"word_count" gorm:"default:0"`
	UniqueWordCount  int    `json:"unique_word_count" gorm:"default:0"`
	DifficultyLevel  string `json:"difficulty_level" gorm:"size:10"` // beginner, intermediate, advanced

	// Reading progress
	LastReadAt      *time.Time `json:"last_read_at"`
	ReadingProgress float64    `json:"reading_progress" gorm:"type:decimal(5,2);default:0.00"` // percentage

	// Content extraction
	ExtractedText string                 `json:"extracted_text" gorm:"type:text"`
	ChapterData   []map[string]interface{} `json:"chapter_data" gorm:"serializer:json"` // [{title, start_pos, end_pos}]

	// Timestamps
	UploadedAt time.Time `json:"uploaded_at" gorm:"autoCreateTime"`

	// Relationships
	ReadingSessions []ReadingSession `json:"reading_sessions,omitempty" gorm:"foreignKey:BookID"`
	Annotations     []BookAnnotation `json:"annotations,omitempty" gorm:"foreignKey:BookID"`
}

// TableName specifies the table name for GORM
func (Book) TableName() string {
	return "books"
}

// ToResponse converts the book to a response format
func (b *Book) ToResponse() BookResponse {
	return BookResponse{
		ID:               b.ID,
		UserID:           b.UserID,
		Title:            b.Title,
		Author:           b.Author,
		Language:         b.Language,
		FileName:         b.FileName,
		FileSize:         b.FileSize,
		MimeType:         b.MimeType,
		ProcessingStatus: b.ProcessingStatus,
		WordCount:        b.WordCount,
		UniqueWordCount:  b.UniqueWordCount,
		DifficultyLevel:  b.DifficultyLevel,
		UploadedAt:       b.UploadedAt,
		LastReadAt:       b.LastReadAt,
		ReadingProgress:  b.ReadingProgress,
		ChapterData:      b.ChapterData,
	}
}

// BookResponse is the response format for book data
type BookResponse struct {
	ID               uint                     `json:"id"`
	UserID           uint                     `json:"user_id"`
	Title            string                   `json:"title"`
	Author           string                   `json:"author"`
	Language         string                   `json:"language"`
	FileName         string                   `json:"file_name"`
	FileSize         int64                    `json:"file_size"`
	MimeType         string                   `json:"mime_type"`
	ProcessingStatus string                   `json:"processing_status"`
	WordCount        int                      `json:"word_count"`
	UniqueWordCount  int                      `json:"unique_word_count"`
	DifficultyLevel  string                   `json:"difficulty_level"`
	UploadedAt       time.Time                `json:"uploaded_at"`
	LastReadAt       *time.Time               `json:"last_read_at"`
	ReadingProgress  float64                  `json:"reading_progress"`
	ChapterData      []map[string]interface{} `json:"chapter_data"`
}

// ReadingSession represents a reading session
type ReadingSession struct {
	ID        uint           `json:"id" gorm:"primarykey"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`

	// Foreign keys
	UserID uint `json:"user_id" gorm:"not null;index"`
	BookID uint `json:"book_id" gorm:"not null;index"`
	User   User `json:"user,omitempty" gorm:"foreignKey:UserID"`
	Book   Book `json:"book,omitempty" gorm:"foreignKey:BookID"`

	// Session timing
	StartTime       time.Time  `json:"start_time" gorm:"autoCreateTime"`
	EndTime         *time.Time `json:"end_time"`
	DurationMinutes int        `json:"duration_minutes"`

	// Progress tracking
	StartPosition int `json:"start_position" gorm:"default:0"` // Character position in text
	EndPosition   int `json:"end_position" gorm:"default:0"`
	WordsLearned  int `json:"words_learned" gorm:"default:0"`
	WordsReviewed int `json:"words_reviewed" gorm:"default:0"`

	// Session data
	NewWordsEncountered []uint `json:"new_words_encountered" gorm:"serializer:json"` // Array of word IDs
	WordsLookedUp       []uint `json:"words_looked_up" gorm:"serializer:json"`       // Array of word IDs
}

// TableName specifies the table name for GORM
func (ReadingSession) TableName() string {
	return "reading_sessions"
}

// BookAnnotation represents an annotation on a book
type BookAnnotation struct {
	ID        uint           `json:"id" gorm:"primarykey"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`

	// Foreign keys
	UserID uint `json:"user_id" gorm:"not null;index"`
	BookID uint `json:"book_id" gorm:"not null;index"`
	User   User `json:"user,omitempty" gorm:"foreignKey:UserID"`
	Book   Book `json:"book,omitempty" gorm:"foreignKey:BookID"`

	// Position in text
	StartPosition int    `json:"start_position" gorm:"not null"`
	EndPosition   int    `json:"end_position" gorm:"not null"`
	SelectedText  string `json:"selected_text" gorm:"type:text;not null"`

	// Annotation data
	AnnotationType string                 `json:"annotation_type" gorm:"size:20;default:word_lookup"` // word_lookup, note, highlight
	AnnotationData map[string]interface{} `json:"annotation_data" gorm:"serializer:json"`             // Flexible data storage

	// Visual styling
	HighlightColor string `json:"highlight_color" gorm:"size:7"` // Hex color code
}

// TableName specifies the table name for GORM
func (BookAnnotation) TableName() string {
	return "book_annotations"
}
